apps/els_lsp/src/els_call_hierarchy_provider.erl

```diff
@@ -0,0 +1,112 @@
+-type state() :: any().
+
+%%==============================================================================
+%% els_provider functions
+%%==============================================================================
+-spec is_enabled() -> boolean().
+is_enabled() -> true.
+
+-spec handle_request(any(), state()) -> {any(), state()}.
+handle_request({prepare, Params}, State) ->
+  {Uri, Line, Char} =
+    els_text_document_position_params:uri_line_character(Params),
+  {ok, Document} = els_utils:lookup_document(Uri),
+  Functions = els_dt_document:wrapping_functions(Document, Line + 1, Char + 1),
+  Items = [function_to_item(Uri, F) ||  F <- Functions],
+  {Items, State};
+handle_request({incoming_calls, Params}, State) ->
+  #{ <<"item">> := #{<<"uri">> := Uri} = Item } = Params,
+  POI = els_call_hierarchy_item:poi(Item),
+  References = els_references_provider:find_references(Uri, POI),
+  Items = [reference_to_item(Reference) || Reference <- References],
+  {incoming_calls(Items), State};
+handle_request({outgoing_calls, Params}, State) ->
+  #{ <<"item">> := Item } = Params,
+  #{ <<"uri">> := Uri } = Item,
+  POI = els_call_hierarchy_item:poi(Item),
+  Applications = applications_in_function_range(Uri, POI),
+  Items = lists:foldl(fun(Application, Acc) ->
+                          case application_to_item(Uri, Application) of
+                            {error, not_found} ->
+                              %% The function may contain a reference
+                              %% to a not yet implemented function
+                              Acc;
+                            {ok, I} ->
+                              [I|Acc]
+                          end
+                      end, [], Applications),
+  {outgoing_calls(lists:reverse(Items)), State}.
+
+%%==============================================================================
+%% Internal functions
+%%==============================================================================
+-spec incoming_calls([els_call_hierarchy_item:item()]) ->
+        [els_call_hierarchy_item:incoming_call()].
+incoming_calls(Items) ->
+  [#{from => Item, fromRanges => [Range]} || #{range := Range} = Item <- Items].
+
+-spec outgoing_calls([els_call_hierarchy_item:item()]) ->
+        [els_call_hierarchy_item:outgoing_call()].
+outgoing_calls(Items) ->
+  [#{to => Item, fromRanges => [Range]} || #{range := Range} = Item <- Items].
+
+-spec function_to_item(uri(), poi()) -> els_call_hierarchy_item:item().
+function_to_item(Uri, Function) ->
+  #{id := Id, range := Range} = Function,
+  Name = els_utils:function_signature(Id),
+  Data = #{poi => Function},
+  els_call_hierarchy_item:new(Name, Uri, Range, Range, Data).
+
+-spec reference_to_item(location()) -> els_call_hierarchy_item:item().
+reference_to_item(Reference) ->
+  #{uri := RefUri, range := RefRange} = Reference,
+  {ok, RefDoc} = els_utils:lookup_document(RefUri),
+  [WrappingPOI] = els_dt_document:wrapping_functions(RefDoc, RefRange),
```

Comment:
can this return more than one POI?

Reply:
In your referenced comment you state zero or one.  When zero, we get a crash.